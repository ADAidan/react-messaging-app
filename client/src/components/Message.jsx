import * as React from "react";
import PropTypes from "prop-types";
import Paper from "@mui/material/Paper";
import Stack from "@mui/material/Stack";
import Typography from "@mui/material/Typography";
import DynamicAvatar from "./DynamicAvatar";

export function MessageContent({ content, bgColor = "#f3f3f3" }) {
  return (
    <Paper
      elevation={2}
      sx={{
        padding: 1,
        marginBottom: 1,
        backgroundColor: bgColor,
      }}
    >
      <Typography
        variant="body2"
        component="p"
        sx={{
          color: bgColor === "#1e90ff" ? "white" : "black",
        }}
      >
        {content}
      </Typography>
    </Paper>
  );
}

export function Message({ messageHeader, children }) {
  return (
    <Stack data-testid="message">
      <Stack paddingX={1}>
        <Stack
          direction="row"
          spacing={1}
          p={1}
          alignItems="start"
          justifyContent="space-between"
        >
          <DynamicAvatar name={messageHeader.author} />
          <Stack
            sx={{
              alignItems: "flex-start",
              justifyContent: "flex-start",
              width: "100%",
            }}
          >
            <Stack
              direction="row"
              spacing={1}
              sx={{
                alignItems: "baseline",
              }}
            >
              <Typography
                variant="subtitle1"
                component="h3"
                data-testid="message-author"
              >
                {messageHeader.author}
              </Typography>
              <Typography
                variant="caption"
                component="p"
                data-testid="message-time"
              >
                {messageHeader.time}
              </Typography>
            </Stack>
            {children}
          </Stack>
        </Stack>
      </Stack>
    </Stack>
  );
}

MessageContent.propTypes = {
  content: PropTypes.string.isRequired,
  bgColor: PropTypes.string.isRequired,
};

Message.propTypes = {
  messageHeader: PropTypes.shape({
    profilePicture: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
  }).isRequired,
  children: PropTypes.node.isRequired,
};
