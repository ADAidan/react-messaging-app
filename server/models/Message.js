const Mongoose = require('mongoose');

const { Schema } = Mongoose;

const MessageSchema = new Schema({
  author: {
    type: Schema.Types.ObjectId,
    ref: 'User',
  },
  content: {
    type: String,
    required: true,
  },
}, {
  timestamps: true,
});

const Message = Mongoose.model('Message', MessageSchema);

module.exports = Message;
