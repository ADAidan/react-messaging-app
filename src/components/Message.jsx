import * as React from 'react';
import PropTypes from 'prop-types';
import Box from '@mui/material/Box';
import Paper from '@mui/material/Paper';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import { AccessTime } from '@mui/icons-material';

function Message({ message }) {
  return (
    <Grid
      data-testid={`message-${message.id}`}
      item
      xs={12}
    >
      <Paper elevation={3}>
        <Box paddingX={1}>
          <Typography variant="subtitle1" component="h3">
            {message.author}
          </Typography>
          <Typography variant="body2" component="p">
            {message.text}
          </Typography>
          <Box sx={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'flex-end',
          }}
          >
            <AccessTime sx={{
              width: 15,
              height: 15,
            }}
            />
            <Typography variant="caption" component="p">
              {message.time}
            </Typography>
          </Box>
        </Box>
      </Paper>
    </Grid>
  );
}

Message.propTypes = {
  message: PropTypes.shape({
    id: PropTypes.number.isRequired,
    author: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
  }).isRequired,
};

export default Message;
