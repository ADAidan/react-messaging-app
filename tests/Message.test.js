import React from 'react';
import renderer from 'react-test-renderer';
import Message from '../src/components/Message';

it('displays the message', () => {
  const message = { 
    id: 1, 
    author: 'AidanDyer', 
    text: 'Hello, World!', 
    time: '10:00AM'
  };

  const component = renderer.create(<Message message={message} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});