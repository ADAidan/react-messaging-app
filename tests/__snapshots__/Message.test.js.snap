// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the message 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-13i4rnv-MuiGrid-root{box-sizing:border-box;margin:0;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;}",
      }
    }
    data-emotion="css 13i4rnv-MuiGrid-root"
  />,
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
    data-testid="message-1"
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-3su884-MuiPaper-root{background-color:#fff;color:rgba(0, 0, 0, 0.87);-webkit-transition:box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;border-radius:4px;box-shadow:0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);}",
        }
      }
      data-emotion="css 3su884-MuiPaper-root"
    />
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-mro3c9{padding-left:8px;padding-right:8px;}",
          }
        }
        data-emotion="css mro3c9"
      />
      <div
        className="MuiBox-root css-mro3c9"
      >
        <style
          dangerouslySetInnerHTML={
            {
              "__html": ".css-10wpov9-MuiTypography-root{margin:0;font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:400;font-size:1rem;line-height:1.75;letter-spacing:0.00938em;}",
            }
          }
          data-emotion="css 10wpov9-MuiTypography-root"
        />
        <h3
          className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
        >
          AidanDyer
        </h3>
        <style
          dangerouslySetInnerHTML={
            {
              "__html": ".css-e784if-MuiTypography-root{margin:0;font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:400;font-size:0.875rem;line-height:1.43;letter-spacing:0.01071em;}",
            }
          }
          data-emotion="css e784if-MuiTypography-root"
        />
        <p
          className="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Hello, World!
        </p>
        <style
          dangerouslySetInnerHTML={
            {
              "__html": ".css-192qrng{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:end;-ms-flex-pack:end;-webkit-justify-content:flex-end;justify-content:flex-end;}",
            }
          }
          data-emotion="css 192qrng"
        />
        <div
          className="MuiBox-root css-192qrng"
        >
          <style
            dangerouslySetInnerHTML={
              {
                "__html": ".css-dbbnwz-MuiSvgIcon-root{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:1em;height:1em;display:inline-block;fill:currentColor;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;-webkit-transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;font-size:1.5rem;width:15px;height:15px;}",
              }
            }
            data-emotion="css dbbnwz-MuiSvgIcon-root"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-dbbnwz-MuiSvgIcon-root"
            data-testid="AccessTimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
            />
            <path
              d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
          <style
            dangerouslySetInnerHTML={
              {
                "__html": ".css-1sn4lm3-MuiTypography-root{margin:0;font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:400;font-size:0.75rem;line-height:1.66;letter-spacing:0.03333em;}",
              }
            }
            data-emotion="css 1sn4lm3-MuiTypography-root"
          />
          <p
            className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
          >
            10:00AM
          </p>
        </div>
      </div>
    </div>
  </div>,
]
`;
